batch/batch/worker/worker.py

```diff
@@ -138,6 +141,78 @@ def free(self, port):
         self.ports.put_nowait(port)
 
 
+class NetworkNamespace:
+    def __init__(self, subnet_index):
+        self.network_name = uuid.uuid4().hex[:5]
+        self.host_ip = f'10.0.{subnet_index}.10'
+        self.job_ip = f'10.0.{subnet_index}.11'
+        self.port = None
+        self.host_port = None
+
+    async def init(self):
+        veth_host = self.network_name + '-host'
+        veth_job = self.network_name + '-job'
+        await check_shell(
+            f'''
+ip netns add {self.network_name} && \
+    ip link add name {veth_host} type veth peer name {veth_job} && \
+    ip link set dev {veth_host} up && \
+    ip link set {veth_job} netns {self.network_name} && \
+    ip address add {self.host_ip}/24 dev {veth_host}
+    ip -n {self.network_name} link set dev {veth_job} up && \
+    ip -n {self.network_name} link set dev lo up && \
+    ip -n {self.network_name} address add {self.job_ip}/24 dev {veth_job} && \
+    ip -n {self.network_name} route add default via {self.host_ip}'''
+        )
+
+        await check_shell(
+            f'''
+iptables -w 10 --append FORWARD --in-interface {veth_host} --jump ACCEPT && \
+    iptables -w 10 --append FORWARD --out-interface {veth_host} --jump ACCEPT && \
+    iptables -w 10 --table nat --append POSTROUTING --source {self.job_ip}/24 --jump MASQUERADE'''
+        )
+
+        # The default resolver in `/etc/resolv.conf` is a local address (127.0.0.53) which is
+        # inaccessible in other net namespaces since they have their own loop device. So we use
+        # google's public resolver.
+        os.makedirs(f'/etc/netns/{self.network_name}')
+        await check_shell(f'echo nameserver 8.8.8.8 | tee /etc/netns/{self.network_name}/resolv.conf')
```

Comment:
This can probably just be done once since we bind mount it to /etc/resolv.conf in the container. This is the "correct" way to set up a resolv.conf for a network namespace but the container might not care about it

Reply:
The reason I write to `/etc/netns/.../resolv.conf` is because of the way network namespaces work on their own. Inside a network namespace, processes will look to `/etc/netns/<namespace_name>/resolv.conf` for a resolver IP, and fall back to `/etc/resolv.conf` if one doesn't exist.

Say we had a namespace `mynet`, we can run ping within that network namespace by `ip netns exec mynet ping google.com`. ping looks for `/etc/netns/mynet/resolv.conf`, can't find it, and falls back to `/etc/resolv.conf` which, since we're only in a separate network namespace, is the same as that on the host. But! Each network namespace has its own loopback interface, and since our host's `/etc/resolv.conf` contains the entry `nameserver 127.0.0.53`, processes in the `mynet` namespace won't see it. The "solution" to this, per the internet, is to use a public resolver and put it in `/etc/netns/mynet/resolv.conf`. I suppose we could do some `iptables` trickery to redirect some made up IP to the resolver on the host, but I'm unsure if that's necessary or wanted.

To actually answer your question, this is great for network namespaces but our containers have their own mount namespace, so they can't see `/etc/netns/mynet/resolv.conf` on the host. But on the flip side they have their own `/etc/resolv.conf` that we can mess with without worry, so the better approach here is just to write our a single `/container_resolv.conf` on the host with the public resolver, and then mount that to `/etc/resolv.conf` in all our containers.

Reply:
Ultimately, I ended up writing separate files to `/etc/netns/.../` anyway because it makes debugging network problems easier. This way you can just ssh into a worker and run `sudo ip netns exec <ns_name> …`